patchwork-extensions-blockentity/src/main/java/net/patchworkmc/mixin/extensions/blockentity/MixinClientPlayNetworkHandler.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Minecraft Forge, Patchwork Project
+ * Copyright (c) 2016-2020, 2019-2020
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.patchworkmc.mixin.extensions.blockentity;
+
+import java.util.Iterator;
+
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.At.Shift;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+import org.spongepowered.asm.mixin.injection.callback.LocalCapture;
+import net.minecraftforge.common.extensions.IForgeTileEntity;
+
+import net.minecraft.block.entity.BlockEntity;
+import net.minecraft.client.network.ClientPlayNetworkHandler;
+import net.minecraft.client.world.ClientWorld;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.packet.s2c.play.BlockEntityUpdateS2CPacket;
+import net.minecraft.network.packet.s2c.play.ChunkDataS2CPacket;
+import net.minecraft.util.math.BlockPos;
+
+import net.fabricmc.fabric.api.block.entity.BlockEntityClientSerializable;
+
+/**
+ * TODO: This Mixin can be less hacky if we can talk to the Fabric API team.
```

Comment:
What needs to be done here?
If this is being done to work around a FAPI patch, could mixin priorities be used to apply over their patches?

Reply:
We need to insert our own hook here. We cannot reuse Fabric's BE sync, because Fabric API requires BE to be sync to implement BlockEntityClientSerializable. We cannot make IForgeTileEntity extends BlockEntityClientSerializable, otherwise every single BlockEntity will implement BlockEntityClientSerializable....

Ideally we should let Fabric's hook checks for BlockEntityClientSerializable, if the BE is not a BlockEntityClientSerializable, then it should call our hook. This part of logic is in Fabric API's mixin, so we cannot mixin it. What I would expect is that, Fabric API moves that logic from the mixin class to a separate normal class, so that we can mixin it. 

Actually everything is working here and we don't need to set priority at all. I chose an Injection point so that our hook will be executed earlier than Fabric API's. We will then check if the BE implements BlockEntityClientSerializable, if that's the case, pass it to Fabric API, otherwise, we handle it and pass it to IForgeTileEntity. Vanilla BEs are handled in onChunkData, so they don't need this kind of callback.